File: pwa/scripts/goals-resolver.js
// goals-resolver.js — Canonical goal normalization for both narrow and rich schema shapes.
// Narrow shape: { calories: 1200, protein: 85, fiber: 25, water_oz: 64 }
// Rich shape:   { calories: { daily: 1200 }, macros: { protein: { floor, target, ceiling } }, ... }
// Returns a resolved object supporting BOTH access patterns.

const Goals = {
  _n(val, def) {
    if (val == null) return def;
    const n = typeof val === 'number' ? val : parseFloat(val);
    return isNaN(n) ? def : n;
  },

  resolve(raw) {
    if (!raw || typeof raw !== 'object') return Goals.resolve({});

    // Calories
    const calories = Goals._n(
      (raw.calories && typeof raw.calories === 'object') ? raw.calories.daily : raw.calories,
      2000
    );

    // Protein: check already-resolved fields first, then rich shape, then narrow
    const proteinTarget = Goals._n(
      raw.proteinTarget
        ?? ((raw.protein && typeof raw.protein === 'object') ? raw.protein.target : null)
        ?? raw.macros?.protein?.target
        ?? raw.macros?.protein?.grams
        ?? (typeof raw.protein === 'number' ? raw.protein : null),
      100
    );
    const proteinFloor = Goals._n(
      raw.proteinFloor
        ?? raw.protein_floor
        ?? ((raw.protein && typeof raw.protein === 'object') ? raw.protein.floor : null)
        ?? raw.macros?.protein?.floor,
      null
    );
    const proteinCeiling = Goals._n(
      raw.proteinCeiling
        ?? raw.protein_ceiling
        ?? ((raw.protein && typeof raw.protein === 'object') ? raw.protein.ceiling : null)
        ?? raw.macros?.protein?.ceiling,
      null
    );

    // Fat
    const fatFloor = Goals._n(
      raw.fatFloor
        ?? ((raw.fat && typeof raw.fat === 'object') ? raw.fat.floor : null)
        ?? raw.macros?.fat?.floor,
      null
    );

    // Fiber
    const fiberTarget = Goals._n(
      raw.fiberTarget
        ?? raw.fiber_g
        ?? ((raw.fiber && typeof raw.fiber === 'object') ? raw.fiber.daily_g : null)
        ?? (typeof raw.fiber === 'number' ? raw.fiber : null),
      25
    );
    const fiberFloor = Goals._n(
      raw.fiberFloor
        ?? raw.fiber_floor_g
        ?? ((raw.fiber && typeof raw.fiber === 'object') ? raw.fiber.floor_g : null),
      null
    );
    const fiberCeiling = Goals._n(
      raw.fiberCeiling
        ?? raw.fiber_ceiling_g
        ?? ((raw.fiber && typeof raw.fiber === 'object') ? raw.fiber.ceiling_g : null),
      null
    );
    const fiberTrackSplit = !!(
      raw.fiberTrackSplit
        ?? raw.fiber_trackSplit
        ?? ((raw.fiber && typeof raw.fiber === 'object') ? raw.fiber.trackSplit : undefined)
    );

    // Water
    const waterTarget = Goals._n(
      raw.waterTarget
        ?? raw.water_oz
        ?? ((raw.water && typeof raw.water === 'object') ? raw.water.daily_oz : null),
      64
    );
    const waterFloor = Goals._n(
      raw.waterFloor
        ?? raw.water_floor_oz
        ?? ((raw.water && typeof raw.water === 'object') ? raw.water.floor_oz : null),
      null
    );

    // Weight
    const weightCurrent = Goals._n(
      raw.weightCurrent
        ?? ((raw.weight && typeof raw.weight === 'object') ? raw.weight.current : null),
      null
    );
    const weightGoal = Goals._n(
      raw.weightGoal
        ?? ((raw.weight && typeof raw.weight === 'object') ? raw.weight.goal : null),
      null
    );
    const weightFloor = Goals._n(
      raw.weightFloor
        ?? ((raw.weight && typeof raw.weight === 'object') ? raw.weight.floor : null),
      null
    );

    // Carbs
    const carbsTracked = !!(
      raw.carbsTracked
        ?? ((raw.carbs && typeof raw.carbs === 'object') ? raw.carbs.tracked : undefined)
    );

    // Pass-through fields
    const bodyComp = raw.bodyComp ?? raw.bodyComposition ?? null;
    const transit = raw.transit ?? null;
    const hardcore = raw.hardcore ?? null;
    const adaptive = raw.adaptive ?? null;
    const timeline = raw.timeline ?? null;

    return {
      // Narrow-shape backward compat (existing consumers keep working)
      calories,
      protein: proteinTarget,
      fiber: fiberTarget,
      water_oz: waterTarget,

      // Rich-shape fields
      proteinTarget,
      proteinFloor,
      proteinCeiling,
      fatFloor,
      fiberFloor,
      fiberCeiling,
      fiberTrackSplit,
      waterFloor,
      waterTarget,
      weightCurrent,
      weightGoal,
      weightFloor,
      carbsTracked,
      bodyComp,
      transit,
      hardcore,
      adaptive,
      timeline,
    };
  },

  // 'low' | 'on_track' | 'high'
  proteinStatus(actual, resolved) {
    const floor = resolved.proteinFloor != null ? resolved.proteinFloor : resolved.proteinTarget;
    const target = resolved.proteinTarget;
    if (actual >= target) return 'high';
    if (actual >= floor) return 'on_track';
    return 'low';
  },

  // 'low' | 'on_track' | 'high'
  fiberStatus(actual, resolved) {
    const floor = resolved.fiberFloor;
    const ceiling = resolved.fiberCeiling;
    if (ceiling != null && actual > ceiling) return 'high';
    if (floor != null && actual < floor) return 'low';
    return 'on_track';
  },

  // 'under' | 'on_track' | 'over'
  caloriesStatus(actual, resolved) {
    const target = resolved.calories;
    if (actual < target - 150) return 'under';
    if (actual > target + 150) return 'over';
    return 'on_track';
  },

  // 'under' | 'on_track' | 'high'
  waterStatus(actualOz, resolved) {
    const floor = resolved.waterFloor;
    const target = resolved.water_oz;
    if (actualOz >= target) return 'high';
    if (floor != null && actualOz < floor) return 'under';
    return 'on_track';
  },
};

window.Goals = Goals;

File: processing/lib/goals_resolver.py
"""
goals_resolver.py — Deterministic goal normalization and status helpers.

Mirrors the JS schema-tolerance logic in score.js / goals.js.
NO LLM calls. Pure Python.

Canonical resolved shape (flat dict):
  calories, protein, fiber, water_oz            — scalars (backward compat)
  protein_floor, protein_ceiling                — scalar or None
  fiber_floor_g, fiber_ceiling_g                — scalar or None
  fiber_track_split                             — bool
  water_floor_oz                                — scalar or None
  weight, body_composition, transit, hardcore   — passthrough dicts or None
"""

from __future__ import annotations


# ---------------------------------------------------------------------------
# Defaults
# ---------------------------------------------------------------------------
_DEFAULT_CALORIES = 2000
_DEFAULT_PROTEIN  = 100
_DEFAULT_FIBER    = 25
_DEFAULT_WATER_OZ = 64


# ---------------------------------------------------------------------------
# resolve
# ---------------------------------------------------------------------------

def resolve(raw: dict) -> dict:
    """Normalize any goals shape to the canonical flat resolved dict.

    Handles:
      - narrow shape: {calories: 850, protein: 85, fiber: 20, water_oz: 64}
      - rich shape:   {calories: {daily: 850}, macros: {protein: {target: 85}}, ...}
      - already-resolved: idempotent, resolve(resolve(x)) == resolve(x)
    """
    if not raw:
        raw = {}

    # --- calories ---
    cal = raw.get("calories")
    if isinstance(cal, dict):
        calories = _num(cal.get("daily") or cal.get("target"), _DEFAULT_CALORIES)
    else:
        calories = _num(cal, _DEFAULT_CALORIES)

    # --- protein ---
    macros = raw.get("macros") or {}
    protein_obj = macros.get("protein") or {}

    protein = _num(
        raw.get("protein")
        or protein_obj.get("target")
        or protein_obj.get("grams"),
        _DEFAULT_PROTEIN,
    )
    protein_floor = _num_or_none(
        raw.get("protein_floor") or protein_obj.get("floor")
    )
    protein_ceiling = _num_or_none(
        raw.get("protein_ceiling") or protein_obj.get("ceiling")
    )

    # --- fiber ---
    fiber_raw = raw.get("fiber")
    if isinstance(fiber_raw, dict):
        fiber       = _num(fiber_raw.get("daily_g") or fiber_raw.get("target"), _DEFAULT_FIBER)
        fiber_floor   = _num_or_none(fiber_raw.get("floor_g"))
        fiber_ceiling = _num_or_none(fiber_raw.get("ceiling_g"))
        fiber_split   = bool(fiber_raw.get("trackSplit") or fiber_raw.get("track_split"))
    else:
        fiber = _num(
            raw.get("fiber_g")
            or (fiber_raw if isinstance(fiber_raw, (int, float)) else None),
            _DEFAULT_FIBER,
        )
        fiber_floor   = _num_or_none(raw.get("fiber_floor_g"))
        fiber_ceiling = _num_or_none(raw.get("fiber_ceiling_g"))
        fiber_split   = bool(raw.get("fiber_track_split") or raw.get("fiber_trackSplit"))

    # --- water ---
    water_raw = raw.get("water")
    if isinstance(water_raw, dict):
        water_oz      = _num(water_raw.get("daily_oz") or water_raw.get("target_oz"), _DEFAULT_WATER_OZ)
        water_floor   = _num_or_none(water_raw.get("floor_oz"))
    else:
        water_oz    = _num(raw.get("water_oz"), _DEFAULT_WATER_OZ)
        water_floor = _num_or_none(raw.get("water_floor_oz"))

    # --- passthrough rich extras ---
    weight           = raw.get("weight") or None
    body_composition = raw.get("bodyComposition") or raw.get("body_composition") or None
    transit          = raw.get("transit") or None
    hardcore         = raw.get("hardcore") or None

    return {
        # backward-compat scalars
        "calories":          calories,
        "protein":           protein,
        "fiber":             fiber,
        "water_oz":          water_oz,
        # extended floor / ceiling
        "protein_floor":     protein_floor,
        "protein_ceiling":   protein_ceiling,
        "fiber_floor_g":     fiber_floor,
        "fiber_ceiling_g":   fiber_ceiling,
        "fiber_track_split": fiber_split,
        "water_floor_oz":    water_floor,
        # rich-shape passthrough
        "weight":            weight,
        "body_composition":  body_composition,
        "transit":           transit,
        "hardcore":          hardcore,
    }


# ---------------------------------------------------------------------------
# Status helpers
# ---------------------------------------------------------------------------

def protein_status(actual: float, resolved: dict) -> str:
    """Return 'low' | 'on_track' | 'high'."""
    target = resolved["protein"]
    floor  = resolved.get("protein_floor") or 0

    if actual >= target:
        return "high"
    if actual >= floor:
        return "on_track"
    return "low"


def fiber_status(actual: float, resolved: dict) -> str:
    """Return 'low' | 'on_track' | 'high'."""
    floor   = resolved.get("fiber_floor_g") or 0
    ceiling = resolved.get("fiber_ceiling_g")  # None = no upper bound

    if ceiling is not None and actual > ceiling:
        return "high"
    if actual >= floor:
        return "on_track"
    return "low"


def calories_status(actual: float, resolved: dict) -> str:
    """Return 'under' | 'on_track' | 'over'."""
    target = resolved["calories"]
    diff   = actual - target

    if diff > 150:
        return "over"
    if diff < -150:
        return "under"
    return "on_track"


def water_status(actual_oz: float, resolved: dict) -> str:
    """Return 'under' | 'on_track' | 'high'."""
    target = resolved["water_oz"]
    floor  = resolved.get("water_floor_oz") or 0

    if actual_oz >= target:
        return "high"
    if actual_oz >= floor:
        return "on_track"
    return "under"


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

def _num(value, default: float) -> float:
    """Return value as float if numeric, else default."""
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return value
    return default


def _num_or_none(value) -> float | None:
    """Return value as float if numeric, else None."""
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return value
    return None
